src/pretty/chain.rs

```diff
@@ -0,0 +1,237 @@
+        // }
+
+        self
+    }
+
+    pub fn print_doc(self, sty: ChainStyle) -> ArenaDoc<'a> {
+        let arena = &self.printer.arena;
+
+        let op_sep = if sty.space_around_op {
+            arena.line()
+        } else {
+            arena.line_()
+        };
+
+        let use_simple_layout = self.chain_len == 1 && sty.no_break_single && !self.has_comment;
+
+        let mut docs = vec![];
+        let mut has_break = false;
+        let mut leading = true;
+        let mut space_after = true;
+        for item in self.items {
+            match item {
+                ChainItem::Body(body) => {
+                    if leading {
+                        docs.push(body);
+                    } else if let Some(last) = docs.last_mut() {
+                        *last += body;
+                    }
+                    leading = false;
+                    space_after = true;
+                }
+                ChainItem::Op(op) => {
+                    if !(has_break && leading || use_simple_layout) {
+                        docs.push(op_sep.clone());
+                    }
+                    has_break = false;
+                    if sty.space_around_op {
+                        docs.push(op + " ");
+                    } else {
+                        docs.push(op);
+                    }
+                    leading = false;
+                    space_after = false;
+                }
+                ChainItem::Comment(cmt) => {
+                    if leading {
+                        docs.push(cmt);
+                    } else if let Some(last) = docs.last_mut() {
+                        *last += if space_after {
+                            arena.space() + cmt
+                        } else {
+                            cmt
+                        }
+                    }
+                    leading = false;
+                    space_after = true;
+                }
+                ChainItem::Attached(cmt) => {
+                    if let Some(last) = docs.last_mut() {
+                        *last += if space_after {
+                            arena.space() + cmt
+                        } else {
+                            cmt
+                        }
+                    }
+                }
+                ChainItem::Linebreak => {
+                    has_break = true;
+                    leading = true;
+                    docs.push(arena.hardline());
+                }
+            }
+        }
+
+        // for doc in &docs {
+        //     println!(":{:?}", doc.pretty(0).to_string());
+        // }
+        // println!();
+
+        let first_doc = docs.remove(0);
+        let follow_docs = arena.concat(docs);
+        if use_simple_layout {
+            (first_doc + follow_docs).group()
+        } else {
+            (first_doc + (follow_docs).nest(2)).group()
+        }
+    }
+}
+
+/// The iterable count equals to the number of chain operators. FIXME
+pub fn iterate_deep_nodes<'a>(
```

Comment:
what does this function do?

Reply:
oh sorry I didn't notice that the docstring wasn't finished.
This is a utility function that iterates through the chain supplied by resolvers.